File: code/ch12/12.1.2.stoneGame.py
from typing import List


class Solution:
    def stoneGame(self, piles: List[int]) -> bool:
        sum = 0
        for i in piles:
            sum += i
        mem = []
        for i in range(len(piles)):
            mem.append([0] * len(piles))

        return 2 * self.find_max(0, len(piles) - 1, piles, mem) > sum

    def find_max(
        self, left: int, right: int, piles: List[int], mem: List[List[int]]
    ) -> int:
        if left < 0 or right < 0 or left > right:
            return 0
        if mem[left][right] != 0:
            return mem[left][right]
        if left == right:
            mem[left][right] = piles[left]
            return piles[left]
        max_stone = max(
            piles[left] + min(
                self.find_max(left + 2, right, piles, mem),
                self.find_max(left + 1, right - 1, piles, mem),
            ),
            piles[right] + min(
                self.find_max(left + 1, right - 1, piles, mem),
                self.find_max(left, right - 2, piles, mem),
            ),  # black formatter added trailing comma
        )
        mem[left][right] = max_stone
        return max_stone

File: code/ch12/12.1.2.StoneGame.cpp
class Solution {
private:
    int find_max(int left, int right, vector<int>& piles, vector<vector<int>>& mem) {
        if (left < 0 || right < 0 || left > right) return 0;

        if (mem[left][right] != 0)
            return mem[left][right];
        
        if (left == right) {
            mem[left][right] = piles[left];
            return piles[left];
        }

        int max_stone = max(piles[left]  + min(find_max(left + 2, right, piles, mem), find_max(left + 1, right - 1, piles, mem)),
                            piles[right] + min(find_max(left + 1, right - 1, piles, mem), find_max(left, right - 2, piles, mem)));
        mem[left][right] = max_stone;
        
        return max_stone;
    }
public:
    bool stoneGame(vector<int>& piles) {
        int sum = 0;

        for (int i = 0; i < piles.size(); i++)
            sum += piles[i];
        
        vector<vector<int>> mem;
        for (int i = 0; i < piles.size(); i++)
            mem.push_back(vector<int>(piles.size(), 0));

        return 2 * find_max(0, piles.size() - 1, piles, mem) > sum;
    }
};
